src/arch/src/aarch64/gic/dist_regs.rs

```diff
@@ -39,66 +39,73 @@ static VGIC_DIST_REGS: &[DistReg] = &[
+    SharedIrq { bits_per_irq: u8 },
+}
+
 /// This is how we represent the registers of the vgic's distributor.
-/// Some of the distributor register (i.e GICD_STATUSR) are simple
-/// registers (i.e they are associated to a 32 bit value).
-/// However, there are other registers that have variable lengths since
-/// they dedicate some of the 32 bits to some specific interrupt. So, their length
-/// depends on the number of interrupts (i.e the ones that are represented as GICD_REG<n>)
-/// in the documentation mentioned above.
 struct DistReg {
     /// Offset from distributor address.
-    base: u32,
-    /// Bits per interrupt.
-    /// Relevant for registers that DO share IRQs.
-    bpi: u8,
-    /// Length of the register.
-    /// Relevant for registers that DO NOT share IRQs.
-    length: u16,
+    offset: u32,
+    /// Distributor register size.
+    size: DistRegSize,
 }
 
 impl DistReg {
-    const fn new(base: u32, bpi: u8, length: u16) -> DistReg {
-        DistReg { base, bpi, length }
+    const fn simple(offset: u32, size: u16) -> DistReg {
+        DistReg {
+            offset,
+            size: DistRegSize::Simple { size },
+        }
     }
 
-    fn compute_reg_range(&self, fd: &DeviceFd) -> Result<(u32, u32)> {
-        // The ARM® TrustZone® implements a protection logic which contains a read-as-zero/write-ignore (RAZ/WI) policy
-        // where:
-        // * A blocked read operation will always return a zero value on the bus, preventing information leak
-        // * A write operation to a forbidden region or peripheral will be ignored
-        // The first part of a shared-irq type of register = (i.e GICD_<REG><0>) is RAZ/WI, so we compute the
-        // base by shifting it with "REG_SIZE * self.bpi" bytes.
-        let base = self.base + REG_SIZE as u32 * self.bpi as u32;
-        let mut end = base;
-        let num_irq = get_num_interrupts(fd)?;
-
-        if self.length > 0 {
-            // This is the single type register (i.e one that is not DIST_X<n>) and for which
-            // the bpi is 0.
-            // Look in the kernel for REGISTER_DESC_WITH_LENGTH.
-            end = base + self.length as u32;
+    const fn shared_irq(offset: u32, bits_per_irq: u8) -> DistReg {
+        DistReg {
+            offset,
+            size: DistRegSize::SharedIrq { bits_per_irq },
         }
-        if self.bpi > 0 {
-            // This is the type of register that takes into account the number of interrupts
-            // that the model has. It is also the type of register where
-            // a register relates to multiple interrupts.
-            end = base + (self.bpi as u32 * (num_irq - super::super::layout::IRQ_BASE) / 8);
-            if self.bpi as u32 * (num_irq - super::super::layout::IRQ_BASE) % 8 > 0 {
-                end += REG_SIZE as u32;
+    }
+}
+
+impl DistReg {
+    fn mem_iter(&self) -> impl Iterator<Item = u32> {
+        let mut start = self.offset;
```

Comment:
Can we assume `bits_per_irq` % 8 is always 0?  Also, can we assume `bits_per_irq` is > 0?

Reply:
For a `DistRegSize::SharedIrq` register, `bits_per_irq` is always > 0. `bits_per_irq % 8` si not always 0.

Reply:
Ok. Shouldn't any register be of `U32_MAX` length, however? I am just trying to understand if we can simplify the logic here, computing the shared irq dist regs based on register size, rather than `bits_per_irq` size, which might result in an offset which is not byte aligned.

Reply:
The size of the registers will vary, but any register should be a multiple of `U32_SIZE`. The problem is that for any shared irq register, you want to avoid the regions below irq 32 and above irq IRQ_MAX, because they are RAZ/WI. That's why we need to compute these values.